File: Android/app/src/main/java/com/biometric/app/sync/ssot/FirebaseSsotSchema.kt
package com.biometric.app.sync.ssot

/**
 * Canonical Firebase Realtime Database contract shared by Web and Android.
 * Firebase is the cross-platform authoritative store. Room is retained only
 * as Android's local/offline cache and queue during migration.
 */
object FirebaseSsotSchema {
    const val ROOT = "owners"
    const val DEFAULT_OWNER_UID = "biometricpayroll"

    val tables: Map<String, String> = linkedMapOf(
        "Employee" to "employees",
        "Shop" to "shops",
        "AttendanceLog" to "attendance",
        "AttendancePunch" to "attendance_punches",
        "SalaryAdvance" to "advance_payments",
        "EmployeeHistory" to "employee_history",
        "CompanyHoliday" to "shop_closed_days",
        "AttendanceRegularization" to "regularizations",
        "LeaveRequest" to "leave_requests",
        "ResignationRequest" to "resignation_requests",
        "SalarySnapshot" to "salary_snapshots",
        "AuditLog" to "audit_logs",
        "DailySummary" to "daily_summaries",
        "ShiftSchedule" to "shift_schedules",
        "PayrollHistory" to "payroll_history",
        "PayrollPreview" to "payroll_previews",
        "PayrollFinalization" to "payroll_finalization",
        "BonusRecord" to "bonus_records",
        "TaxDeclaration" to "tax_declarations",
        "FBPComponent" to "fbp_components",
        "FlexibleBenefitDeclaration" to "fbp_declarations",
        "FeatureSettings" to "feature_settings",
        "CompanySetting" to "company_settings",
        "ProfessionalTaxSlab" to "professional_tax_slabs",
        "YearEndSummary" to "year_end_summaries",
        "FnFSettlement" to "fnf_settlements",
        "ReportDefinition" to "report_definitions",
        "GeoPunchAudit" to "geo_punch_audits"
    )

    fun tablePath(ownerUid: String, table: String): String =
        "$ROOT/${escape(ownerUid)}/${escape(table)}"

    fun recordPath(ownerUid: String, table: String, recordId: String): String =
        "${tablePath(ownerUid, table)}/${escape(recordId)}"

    fun isTable(table: String?): Boolean =
        !table.isNullOrBlank() && tables.values.any { it == table.trim() }

    private fun escape(value: String): String = value.trim()
        .replace(".", "_dot_")
        .replace("#", "_hash_")
        .replace("$", "_dollar_")
        .replace("[", "_open_")
        .replace("]", "_close_")
        .replace("/", "_slash_")
}

File: Web/Payroll.Shared/Firebase/FirebaseSsotSchema.cs
namespace Payroll.Shared.Firebase;

/// <summary>
/// Canonical Firebase Realtime Database contract for the payroll application.
/// Firebase is the only cross-platform authoritative data store. Web EF/SQL
/// and Android Room are compatibility/cache layers during the controlled
/// migration and must never become competing sources of truth.
/// </summary>
public static class FirebaseSsotSchema
{
    public const string Root = "owners";
    public const string DefaultOwnerUid = "biometricpayroll";

    public static readonly IReadOnlyDictionary<string, string> Tables =
        new Dictionary<string, string>(StringComparer.Ordinal)
        {
            ["Employee"] = "employees",
            ["Shop"] = "shops",
            ["AttendanceLog"] = "attendance",
            ["AttendancePunch"] = "attendance_punches",
            ["SalaryAdvance"] = "advance_payments",
            ["EmployeeHistory"] = "employee_history",
            ["CompanyHoliday"] = "shop_closed_days",
            ["AttendanceRegularization"] = "regularizations",
            ["LeaveRequest"] = "leave_requests",
            ["ResignationRequest"] = "resignation_requests",
            ["SalarySnapshot"] = "salary_snapshots",
            ["AuditLog"] = "audit_logs",
            ["DailySummary"] = "daily_summaries",
            ["ShiftSchedule"] = "shift_schedules",
            ["PayrollHistory"] = "payroll_history",
            ["PayrollPreview"] = "payroll_previews",
            ["PayrollFinalization"] = "payroll_finalization",
            ["BonusRecord"] = "bonus_records",
            ["TaxDeclaration"] = "tax_declarations",
            ["FBPComponent"] = "fbp_components",
            ["FlexibleBenefitDeclaration"] = "fbp_declarations",
            ["FeatureSettings"] = "feature_settings",
            ["CompanySetting"] = "company_settings",
            ["ProfessionalTaxSlab"] = "professional_tax_slabs",
            ["YearEndSummary"] = "year_end_summaries",
            ["FnFSettlement"] = "fnf_settlements",
            ["ReportDefinition"] = "report_definitions",
            ["GeoPunchAudit"] = "geo_punch_audits",
            ["EmployeePresence"] = "presence"
        };

    /// <summary>Firebase owner-scoped path for a table.</summary>
    public static string TablePath(string ownerUid, string table)
        => $"{Root}/{Escape(ownerUid)}/{Escape(table)}";

    /// <summary>Firebase owner-scoped path for a single record.</summary>
    public static string RecordPath(string ownerUid, string table, string recordId)
        => $"{TablePath(ownerUid, table)}/{Escape(recordId)}";

    public static bool IsTable(string table)
        => !string.IsNullOrWhiteSpace(table) && Tables.Values.Contains(table.Trim(), StringComparer.Ordinal);

    public static string? EntityForTable(string table)
    {
        if (string.IsNullOrWhiteSpace(table)) return null;
        return Tables.FirstOrDefault(x => string.Equals(x.Value, table.Trim(), StringComparison.Ordinal)).Key;
    }

    private static string Escape(string value)
        => (value ?? string.Empty).Trim()
            .Replace(".", "_dot_", StringComparison.Ordinal)
            .Replace("#", "_hash_", StringComparison.Ordinal)
            .Replace("$", "_dollar_", StringComparison.Ordinal)
            .Replace("[", "_open_", StringComparison.Ordinal)
            .Replace("]", "_close_", StringComparison.Ordinal)
            .Replace("/", "_slash_", StringComparison.Ordinal);
}
